backend/src/schema/resolvers/emails.spec.js

```diff
@@ -0,0 +1,298 @@
+import Factory from '../../seed/factories'
+import { gql } from '../../jest/helpers'
+import { getDriver, neode as getNeode } from '../../bootstrap/neo4j'
+import createServer from '../../server'
+import { createTestClient } from 'apollo-server-testing'
+
+const factory = Factory()
+const neode = getNeode()
+
+let mutate
+let authenticatedUser
+let user
+let variables
+const driver = getDriver()
+
+beforeEach(async () => {
+  variables = {}
+})
+
+beforeAll(() => {
+  const { server } = createServer({
+    context: () => {
+      return {
+        driver,
+        neode,
+        user: authenticatedUser,
+      }
+    },
+  })
+  mutate = createTestClient(server).mutate
+})
+
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('AddEmailAddress', () => {
+  const mutation = gql`
+    mutation($email: String!) {
+      AddEmailAddress(email: $email) {
+        email
+        verifiedAt
+        createdAt
+      }
+    }
+  `
+  beforeEach(() => {
+    variables = { ...variables, email: 'new-email@example.org' }
+  })
+
+  describe('unauthenticated', () => {
+    beforeEach(() => {
+      authenticatedUser = null
+    })
+
+    it('throws AuthorizationError', async () => {
+      await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+        data: { AddEmailAddress: null },
+        errors: [{ message: 'Not Authorised!' }],
+      })
+    })
+  })
+
+  describe('authenticated', () => {
+    beforeEach(async () => {
+      user = await factory.create('User', { id: '567', email: 'user@example.org' })
+      authenticatedUser = await user.toJson()
+    })
+
+    describe('email attribute is not a valid email', () => {
+      beforeEach(() => {
+        variables = { ...variables, email: 'foobar' }
+      })
+
+      it('throws UserInputError', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: { AddEmailAddress: null },
+          errors: [{ message: 'must be a valid email' }],
+        })
+      })
+    })
+
+    describe('email attribute is a valid email', () => {
+      it('creates a new unverified `EmailAddress` node', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: {
+            AddEmailAddress: {
+              email: 'new-email@example.org',
+              verifiedAt: null,
+              createdAt: expect.any(String),
+            },
+          },
+          errors: undefined,
+        })
+      })
+
+      it('connects `UnverifiedEmailAddress` to the authenticated user', async () => {
+        await mutate({ mutation, variables })
+        const result = await neode.cypher(`
+        MATCH(u:User)-[:PRIMARY_EMAIL]->(:EmailAddress {email: "user@example.org"})
+        MATCH(u:User)<-[:BELONGS_TO]-(e:UnverifiedEmailAddress {email: "new-email@example.org"})
+        RETURN e
+      `)
+        const email = neode.hydrateFirst(result, 'e', neode.model('UnverifiedEmailAddress'))
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T09:18:17Z" title="Wednesday, October 2nd 2019, 11:18:17 am +02:00">Oct 2, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,298 @@
+import Factory from '../../seed/factories'
+import { gql } from '../../jest/helpers'
+import { getDriver, neode as getNeode } from '../../bootstrap/neo4j'
+import createServer from '../../server'
+import { createTestClient } from 'apollo-server-testing'
+
+const factory = Factory()
+const neode = getNeode()
+
+let mutate
+let authenticatedUser
+let user
+let variables
+const driver = getDriver()
+
+beforeEach(async () => {
+  variables = {}
+})
+
+beforeAll(() => {
+  const { server } = createServer({
+    context: () => {
+      return {
+        driver,
+        neode,
+        user: authenticatedUser,
+      }
+    },
+  })
+  mutate = createTestClient(server).mutate
+})
+
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('AddEmailAddress', () => {
+  const mutation = gql`
+    mutation($email: String!) {
+      AddEmailAddress(email: $email) {
+        email
+        verifiedAt
+        createdAt
+      }
+    }
+  `
+  beforeEach(() => {
+    variables = { ...variables, email: 'new-email@example.org' }
+  })
+
+  describe('unauthenticated', () => {
+    beforeEach(() => {
+      authenticatedUser = null
+    })
+
+    it('throws AuthorizationError', async () => {
+      await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+        data: { AddEmailAddress: null },
+        errors: [{ message: 'Not Authorised!' }],
+      })
+    })
+  })
+
+  describe('authenticated', () => {
+    beforeEach(async () => {
+      user = await factory.create('User', { id: '567', email: 'user@example.org' })
+      authenticatedUser = await user.toJson()
+    })
+
+    describe('email attribute is not a valid email', () => {
+      beforeEach(() => {
+        variables = { ...variables, email: 'foobar' }
+      })
+
+      it('throws UserInputError', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: { AddEmailAddress: null },
+          errors: [{ message: 'must be a valid email' }],
+        })
+      })
+    })
+
+    describe('email attribute is a valid email', () => {
+      it('creates a new unverified `EmailAddress` node', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: {
+            AddEmailAddress: {
+              email: 'new-email@example.org',
+              verifiedAt: null,
+              createdAt: expect.any(String),
+            },
+          },
+          errors: undefined,
+        })
+      })
+
+      it('connects `UnverifiedEmailAddress` to the authenticated user', async () => {
+        await mutate({ mutation, variables })
+        const result = await neode.cypher(`
+        MATCH(u:User)-[:PRIMARY_EMAIL]->(:EmailAddress {email: "user@example.org"})
+        MATCH(u:User)<-[:BELONGS_TO]-(e:UnverifiedEmailAddress {email: "new-email@example.org"})
+        RETURN e
+      `)
+        const email = neode.hydrateFirst(result, 'e', neode.model('UnverifiedEmailAddress'))
```


I'll have to read up on this method... `hydrateFirst` sounds like some advice you give someone before going out drinking :laughing:

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-02T13:24:16Z" title="Wednesday, October 2nd 2019, 3:24:16 pm +02:00">Oct 2, 2019</time>_

---


:laughing: it's the `neode` way of parsing a transaction record into a `neode` node. It needs to know the `model` name and the variable in the cypher statement.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T14:29:32Z" title="Wednesday, October 2nd 2019, 4:29:32 pm +02:00">Oct 2, 2019</time>_

---


great! thanks for that... I'll still have a look at the docs, and ideally use it myself... I guessed that is what the `'e'` was, but I would have just used `email` personally... I don't know what the 4 characters less buys you, but I know programmers are different.